cpp/src/arrow/util/fixed_width_internal.h

```diff
@@ -0,0 +1,364 @@
+/// `DataType::byte_width()`. `DataType::byte_width()` will return 0 for
+/// BOOL, while this function will return `-1`. This is done because 0 is
+/// a valid return value for FIXED_SIZE_LIST with size 0 or `FIXED_SIZE_BINARY` with
+/// size 0.
+///
+/// \pre The instance of the array where this type is from must pass
+///      `IsFixedWidthLike(array)` and should not be BOOL.
+/// \return The fixed-byte width of the values or -1 if the type is BOOL or not
+///         fixed-width like. 0 is a valid return value as fixed-size-lists
+///         and fixed-size-binary with size 0 are allowed.
+inline int64_t FixedWidthInBytes(const DataType& type) {
+  auto type_id = type.id();
+  if (is_fixed_width(type_id)) {
+    const int32_t num_bits = type.bit_width();
+    return (type_id == Type::BOOL) ? -1 : num_bits / 8;
+  }
+  if (type_id == Type::FIXED_SIZE_LIST) {
+    auto& fsl = ::arrow::internal::checked_cast<const FixedSizeListType&>(type);
+    return internal::FixedWidthInBytesFallback(fsl);
+  }
+  return -1;
+}
+
+/// \brief Get the fixed-width in bits of a type if it is a fixed-width like
+/// type.
+///
+/// \return The bit-width of the values or -1
+/// \see FixedWidthInBytes
+inline int64_t FixedWidthInBits(const DataType& type) {
+  auto type_id = type.id();
+  if (is_fixed_width(type_id)) {
+    return type.bit_width();
+  }
+  const int64_t byte_width = FixedWidthInBytes(type);
+  if (ARROW_PREDICT_FALSE(byte_width < 0)) {
+    return -1;
+  }
+  return byte_width * 8;
+}
+
+namespace internal {
+
+/// \brief Allocate an ArrayData for a type that is fixed-width like.
+///
+/// This function performs the same checks performed by
+/// `IsFixedWidthLike(source, false)`. If `source.type` is not a simple
+/// fixed-width type, caller should make sure it passes the
+/// `IsFixedWidthLike(source)` checks. That guarantees that it's possible to
+/// allocate an array that can serve as a destination for a kernel that writes values
+/// through a single pointer to fixed-width byte blocks.
+///
+/// \param[in] length The length of the array to allocate (unrelated to the length of
+///                   the source array)
+/// \param[in] source The source array that carries the type information and the
+///                   validity bitmaps that are relevant for the type validation
+///                   when the source is a FixedSizeList.
+/// \see IsFixedWidthLike
+ARROW_EXPORT Status PreallocateFixedWidthArrayData(::arrow::compute::KernelContext* ctx,
+                                                   int64_t length,
+                                                   const ArraySpan& source,
+                                                   bool allocate_validity,
+                                                   ArrayData* out);
+
+/// \pre same as OffsetPointerOfFixedWidthValues
+/// \pre source.type->id() != Type::BOOL
+ARROW_EXPORT const uint8_t* OffsetPointerOfFixedWidthValuesFallback(
```

Comment:
What's the intention behind the `*Fallback` functions? Are there plans to introduce faster paths later or are they just meant to hide implementation details?

Reply:
It's to allow the hot code path to be inlined while the cold code stays behind the function call.

```cpp
inline const uint8_t* OffsetPointerOfFixedWidthValues(const ArraySpan& source) {
  auto type_id = source.type->id();
  if (ARROW_PREDICT_TRUE(is_fixed_width(type_id))) {
    if (ARROW_PREDICT_FALSE(type_id == Type::BOOL)) {
      // BOOL arrays are bit-packed, thus a byte-aligned pointer cannot be produced in the
      // general case. Returning something for BOOL arrays that happen to byte-align
      // because offset=0 would create too much confusion.
      return nullptr;
    }
    return source.GetValues<uint8_t>(1, 0) + source.offset * source.type->byte_width();
  }
  return internal::OffsetPointerOfFixedWidthValuesFallback(source);
}
```

This is good for binary size while still allowing code like

```cpp
if (is_fixed_width(array.type->id())) {
  auto ptr = OffsetPointerOfFixedWidthValues(array);
  ptr->...
}
```

to lower to the equivalent of

```cpp
return source.GetValues<uint8_t>(1, 0) + source.offset * source.type->byte_width();
```

The binary size savings add up because the number of templates specializations using these functions is a proportional to `O(number of small powers-of-2 * index types * ...)` and replacing a `GetValues` call with `OffsetPointerOfFixedWidthValues` costs nothing for the common fixed-width types.

Reply:
Sounds good! Thanks for the explanation.

Reply:
For instance, loop I just wrote:

```cpp
  for (size_t i = 0; i < num_chunks; ++i) {
    const ArraySpan chunk{*chunks[i]->data()};
    if (chunk.type->id() == Type::BOOL) {
      src_chunks_[i] = chunk.GetValues<uint8_t>(1, 0);
      src_offsets_[i] = chunk.offset;
    } else {
      src_chunks_[i] = util::OffsetPointerOfFixedWidthValues(chunk);
      src_offsets_[i] = 0;
    }
  }
```

Reply:
I'm not inlining the entire function, just the cases where it lowers to a simple `GetValues` and exposing the conditionals so the compiler can avoid performing redundant checks where possible.

I understand it looks obsessive, but do you want me to put effort to undo this work?

Keeping the top-level checks in the header also serves as documentation.

```
const uint8_t* OffsetPointerOfFixedWidthValues(const ArraySpan& source) {
  auto type_id = source.type->id();
  if (ARROW_PREDICT_TRUE(is_fixed_width(type_id))) {
    if (ARROW_PREDICT_FALSE(type_id == Type::BOOL)) {
      // BOOL arrays are bit-packed, thus a byte-aligned pointer cannot be produced in the
      // general case. Returning something for BOOL arrays that happen to byte-align
      // because offset=0 would create too much confusion.
      return nullptr;
    }
    return source.GetValues<uint8_t>(1, 0) + source.offset * source.type->byte_width();
  }
  return internal::OffsetPointerOfFixedWidthValuesFallback(source);
}
```

This is how I use it in a loop

```cpp
  src_chunks_.resize(chunks.size());
  for (size_t i = 0; i < chunks.size(); ++i) {
    const ArraySpan chunk{*chunks[i]->data()};
    DCHECK(util::IsFixedWidthLike(chunk));
    if (chunk.type->id() == Type::BOOL) {
      src_chunks_[i] = chunk.GetValues<uint8_t>(1, 0);
    } else {
      src_chunks_[i] = util::OffsetPointerOfFixedWidthValues(chunk);
    }
  }
```

Reply:
I pushed a commit removing all the inline code from the `.h` except for the template function.